core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -81,23 +85,60 @@ private[coordinator] object TransactionMetadata {
       case unknown => throw new IllegalStateException("Unknown transaction state byte " + unknown + " from the transaction status message")
     }
   }
+
+  private val validPreviousStates: Map[TransactionState, Set[TransactionState]] =
+    Map(Empty -> Set(),
+      Ongoing -> Set(Ongoing, Empty, CompleteCommit, CompleteAbort),
+      PrepareCommit -> Set(PrepareCommit, Ongoing),
+      PrepareAbort -> Set(PrepareAbort, Ongoing),
+      CompleteCommit -> Set(CompleteCommit, PrepareCommit),
+      CompleteAbort -> Set(CompleteAbort, PrepareAbort))
 }
 
 @nonthreadsafe
 private[coordinator] class TransactionMetadata(val pid: Long,
                                                var epoch: Short,
                                                val txnTimeoutMs: Int,
-                                               var state: TransactionState) {
-
-  def this(pid: Long, epoch: Short, txnTimeoutMs: Int) = this(pid, epoch, txnTimeoutMs, Empty)
+                                               var state: TransactionState,
+                                               val topicPartitions: mutable.Set[TopicPartition]) {
 
-  // participated partitions in this transaction
-  val topicPartitions = mutable.Set.empty[TopicPartition]
+  // pending state is used to indicate the state that this transaction is going to
+  // transit to, and for blocking future attempts to transit it again if it is not legal;
+  // initialized as the same as the current state
+  private var pendingState = state
 
-  def addPartitions(partitions: Set[TopicPartition]): Unit = {
+  def addPartitions(partitions: collection.Set[TopicPartition]): Unit = {
     topicPartitions ++= partitions
   }
 
+  def abortPendingTransition(): TransactionState = {
+    val abortedState = pendingState
+    pendingState = state
```

Comment:
It might be clearer to make `pendingState` an `Option` and set it to `None` here.